ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/threadext.ml

```diff
@@ -55,69 +55,29 @@ let thread_iter f xs =
   match thread_iter_all_exns f xs with [] -> () | (_, e) :: _ -> raise e
 
 module Delay = struct
-  (* Concrete type is the ends of a pipe *)
-  type t = {
-      (* A pipe is used to wake up a thread blocked in wait: *)
-      mutable pipe_in: Unix.file_descr option
-    ; (* Indicates that a signal arrived before a wait: *)
-      mutable signalled: bool
-    ; m: M.t
-  }
+  type t
 
-  let make () = {pipe_in= None; signalled= false; m= M.create ()}
+  external make : unit -> t = "caml_xapi_delay_create"
 
-  exception Pre_signalled
+  external signal : t -> unit = "caml_xapi_delay_signal"
 
-  let wait (x : t) (seconds : float) =
-    let to_close = ref [] in
-    let close' fd =
-      if List.mem fd !to_close then Unix.close fd ;
-      to_close := List.filter (fun x -> fd <> x) !to_close
-    in
-    finally
-      (fun () ->
-        try
-          let pipe_out =
-            Mutex.execute x.m (fun () ->
-                if x.signalled then (
-                  x.signalled <- false ;
-                  raise Pre_signalled
-                ) ;
-                let pipe_out, pipe_in = Unix.pipe () in
-                (* these will be unconditionally closed on exit *)
-                to_close := [pipe_out; pipe_in] ;
-                x.pipe_in <- Some pipe_in ;
-                x.signalled <- false ;
-                pipe_out
-            )
-          in
-          let open Xapi_stdext_unix.Unixext in
-          (* flush the single byte from the pipe *)
-          try
-            let (_ : string) =
-              time_limited_single_read pipe_out 1 ~max_wait:seconds
-            in
-            false
-          with Timeout -> true
-          (* return true if we waited the full length of time, false if we were woken *)
-        with Pre_signalled -> false
-      )
-      (fun () ->
-        Mutex.execute x.m (fun () ->
-            x.pipe_in <- None ;
-            List.iter close' !to_close
-        )
-      )
+  external wait : t -> int64 -> bool = "caml_xapi_delay_wait"
 
-  let signal (x : t) =
-    Mutex.execute x.m (fun () ->
-        match x.pipe_in with
-        | Some fd ->
-            ignore (Unix.write fd (Bytes.of_string "X") 0 1)
-        | None ->
-            x.signalled <- true
-        (* If the wait hasn't happened yet then store up the signal *)
-    )
+  let wait d t =
+    if t <= 0. then
+      true
+    else
+      match Mtime.Span.of_float_ns (t *. 1e9) with
+      | Some span -> (
+          let now = Mtime_clock.now () in
+          match Mtime.add_span now span with
+          | Some deadline ->
+              wait d (Mtime.to_uint64_ns deadline)
+          | None ->
+              invalid_arg "Time specified too big"
+        )
+      | None ->
+          invalid_arg "Time specified too big"
```

Comment:
The case where the result will overflow is exceedingly rare, I'd rather clamp the value and avoid the exception:

```suggestion
      | Some span ->
          let now = Mtime_clock.now () in
          let deadline =
            Mtime.add_span now span |> Option.value ~default:Mtime.max_stamp
          in
          wait d (Mtime.to_uint64_ns deadline)
      | None ->
          invalid_arg "Time specified too big"
```

Reply:
Sounds correct. Note that the surrounding `match` (`match Mtime.Span.of_float_ns (t *. 1e9) with`) also handles a similar case, should not be changed also in a similar way?
Can you suggest a code that handle both having a single final `wait`?

Reply:
`2^53` seconds is more than 285 million years, the "already" seems a bit preposterous to me. But fine with the change, I'll update

Reply:
Sorry, that's nanoseconds, as the function name implies. (~104 days)
https://ocaml.org/p/mtime/latest/doc/Mtime/Span/index.html

Reply:
"merged"